### DISCUSSION TITLE
Help needed with fetch-based networking

### DISCUSSION THREAD

**Discussion by chschnell** [copy/v86#1180] [Q&A] [answered]
In the [documentation](https://github.com/copy/v86/blob/master/docs/networking.md#fetch-based-networking) about fetch-based networking it says:

> HTTP requests are translated into calls to fetch (which only works on CORS-enabled hosts)

What is meant by "CORS-enabled hosts"? Do you mean HTTP servers that allow/deny requests based on HTTP CORS headers (I figure since it also says that "no external relay is used")?

Is it at all possible to use an external relay with fetch-based networking? If it is, how could I run a local relay, preferrably on linux?

Thanks!

**Comment by SuperMaxusa** [comment, marked as answer]
> What is meant by "CORS-enabled hosts"?

Each HTTP server can control whether the client from client's origin host (with this method and headers) can load data from that server or not. "CORS-enabled" host meant that server allows your origin host, headers and method that you use. 

For example, how this looks on server's response and client's request:

```http
// client

GET / HTTP/1.1
Host: example.com
Origin: http://someweb.site
```

```http
// server that allowed CORS

HTTP/1.1 200 OK
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: *
Access-Control-Allow-Headers: *
```
Server must include these headers or CORS was not permitted.

> Is it at all possible to use an external relay with fetch-based networking?

If you about relay for bypassing CORS, you can replace this line:
https://github.com/copy/v86/blob/ba7f9405840b6ae476bf01a5b4dfb7f1a6e48055/src/browser/starter.js#L316
to 
```js
this.network_adapter = new FetchNetworkAdapter(this.bus, { cors_proxy: "<http://your-cors-proxy>" });
```

> If it is, how could I run a local relay, preferrably on linux?

Look at https://github.com/Rob--W/cors-anywhere

For interest, I wrote a simple CORS proxy for Deno that packaged as a one-liner (uses port 8000 by default):
```
deno eval "Deno.serve(async r=>{let f = await fetch(new URL(r.url).searchParams.get('url'),r);let h = new Headers(f.headers);h.set('Access-Control-Allow-Origin','*');return new Response(await f.arrayBuffer(),{headers:h})})"
```
Template for `cors_proxy`: `http://localhost:8000/?url=`

**Reply by chschnell** [reply]
> Look at https://github.com/Rob--W/cors-anywhere

Yes, thank you so much, this is exactly what I meant!

I said "relay" because the idea of a reverse proxy that adds CORS headers didn't come to me.

I will try it out.

**Reply by chschnell** [reply]
> For interest, I wrote a simple CORS proxy for Deno that packaged as a one-liner (uses port 8000 by default):

Nice!

Inspired by that, I've created a simple CORS proxy in PHP, see **[uncors](https://github.com/chschnell/uncors)**.

It's not a one-liner like yours, but straightforward if you already have a local web server running that supports PHP.

**Reply by SuperMaxusa** [reply]
> Inspired by that, I've created a simple CORS proxy in PHP, see **[uncors](https://github.com/chschnell/uncors)**.

Cool, I think that this will come in handy on my local server.
